echo-pubsub/src/test/groovy/com/netflix/spinnaker/echo/pubsub/amazon/AmazonSQSSubscriberSpec.groovy

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2018 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.echo.pubsub.amazon
+
+import com.amazonaws.services.sns.AmazonSNS
+import com.amazonaws.services.sqs.AmazonSQS
+import com.amazonaws.services.sqs.model.CreateQueueResult
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.netflix.spectator.api.Registry
+import com.netflix.spinnaker.echo.config.amazon.AmazonPubsubProperties
+import com.netflix.spinnaker.echo.pubsub.PubsubMessageHandler
+import spock.lang.Shared
+import spock.lang.Specification
+import spock.lang.Subject
+
+class AmazonSQSSubscriberSpec extends Specification {
+
+  AmazonSNS amazonSNS = Mock()
+  AmazonSQS amazonSQS = Mock()
+  PubsubMessageHandler pubsubMessageHandler = Mock()
+  Registry registry = Mock()
+
+  ARN queueARN = new ARN("arn:aws:sqs:us-west-2:100:queueName")
+  ARN topicARN = new ARN("arn:aws:sns:us-west-2:100:topicName")
+  AmazonPubsubProperties.AmazonPubsubSubscription subscription =
+    new AmazonPubsubProperties.AmazonPubsubSubscription('aws_events', 'mgmt', topicARN.arn, queueARN.arn, "", null)
+
+  @Shared
+  def objectMapper = new ObjectMapper()
+
+  @Subject
+  def subject = new SQSSubscriber(
+    objectMapper,
+    subscription,
+    pubsubMessageHandler,
+    amazonSNS,
+    amazonSQS,
+    registry)
+
+  void "should create queue if it does not exist"() {
+    when:
+    def queueId = SQSSubscriber.ensureQueueExists(amazonSQS, queueARN, topicARN, 1)
+
+    then:
+    queueId == "my-queue-url"
+
+    1 * amazonSQS.createQueue(queueARN.name) >> { new CreateQueueResult().withQueueUrl("my-queue-url") }
+    1 * amazonSQS.setQueueAttributes("my-queue-url", [
+      "Policy": SQSSubscriber.buildSQSPolicy(queueARN, topicARN).toJson(),
+      "MessageRetentionPeriod": "1"
+    ])
+    0 * _
+  }
+
+  def 'should unmarshall an sns notification message'() {
+    given:
+    String payload = "{\"Records\":[{" +
```

Comment:
Not sure how tedious it was to create this ... but you could use a multiline string which might look a little better

"""
{"Records": [{ ... }]}
"""